File: backend/app/models/style_memory.py
"""Style Memory API models (spec §12).

Everything the user can SEE, CORRECT and RESET about what WTM has learned. The
shape is deliberately explicit rather than an opaque blob: each preference
carries its own confidence and whether the user stated it, because the UI has to
be able to tell "you told us this" apart from "we noticed this" (§12.3).
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.services.style_memory import FACETS, REJECTION_REASONS

RejectionReason = Literal[
    "identity_issue",
    "garment_issue",
    "not_my_style",
    "body_proportion_issue",
    "color_issue",
    "occasion_mismatch",
    "other",
]

Facet = Literal[
    "preferred_colors",
    "avoided_colors",
    "preferred_silhouettes",
    "avoided_silhouettes",
    "preferred_aesthetics",
    "preferred_occasions",
    "preferred_moods",
    "fit_visual_preferences",
]

# Guardrail: the Literal above and the service vocabulary must never drift.
assert set(FACETS) == set(Facet.__args__)  # type: ignore[attr-defined]
assert set(REJECTION_REASONS) == set(RejectionReason.__args__)  # type: ignore[attr-defined]


class PreferenceItem(BaseModel):
    value: str
    weight: float
    #: 0..1. Below ~0.35 the app must phrase this as a hunch, never a fact.
    confidence: float
    #: 'stated' = the user said so. 'inferred' = we noticed it.
    source: str
    updated_at: str


class StyleMemoryProfile(BaseModel):
    """The whole of what WTM believes about one user's taste."""

    version: int = 1
    confidence: float = 0
    signal_count: int = 0
    personalization_enabled: bool = True
    #: One hedged sentence, or null when we do not know enough to say anything.
    preference_summary: str | None = None
    preferred_colors: list[PreferenceItem] = Field(default_factory=list)
    avoided_colors: list[PreferenceItem] = Field(default_factory=list)
    preferred_silhouettes: list[PreferenceItem] = Field(default_factory=list)
    avoided_silhouettes: list[PreferenceItem] = Field(default_factory=list)
    preferred_aesthetics: list[PreferenceItem] = Field(default_factory=list)
    preferred_occasions: list[PreferenceItem] = Field(default_factory=list)
    preferred_moods: list[PreferenceItem] = Field(default_factory=list)
    fit_visual_preferences: list[PreferenceItem] = Field(default_factory=list)


class StyleMemorySignalIn(BaseModel):
    """A taste signal the app observed. `dedupe_key` makes a retry a no-op."""

    signal_type: Literal[
        "keep_look",
        "reject_look",
        "save_product",
        "unsave_product",
        "wear_again",
        "mood_selected",
        "occasion_selected",
        "share_look",
        "save_look",
        "manual_preference",
        "event_planned",
    ]
    entity_type: str | None = Field(default=None, max_length=40)
    entity_id: str | None = Field(default=None, max_length=120)
    value: str | None = Field(default=None, max_length=80)
    reason: RejectionReason | None = None
    #: Descriptive attributes the app already knows (mood, occasion). Colours and
    #: silhouettes for a rendered look are resolved SERVER-side from the user's
    #: own wardrobe rows, never accepted from the client.
    mood: str | None = Field(default=None, max_length=40)
    occasion: str | None = Field(default=None, max_length=40)
    dedupe_key: str | None = Field(default=None, max_length=120)


class StyleMemorySignalResponse(BaseModel):
    #: False when the dedupe key had already been recorded — an honest no-op.
    recorded: bool
    profile: StyleMemoryProfile
    #: The restrained "WTM learned…" line, or null when nothing new is sayable.
    learned: str | None = None


class StyleMemoryCorrection(BaseModel):
    facet: Facet
    value: str = Field(min_length=1, max_length=80)
    #: True removes this single entry instead of stating it.
    remove: bool = False


class PersonalizationToggle(BaseModel):
    enabled: bool


class StyleMemoryResetResponse(BaseModel):
    deleted_signals: int
    profile: StyleMemoryProfile


class TryOnFeedback(BaseModel):
    """The verdict on one render (§18).

    A rejection is a taste signal, NOT a refund request: technical and objective
    quality failures are already refunded by the worker before the user ever
    sees a result (§19.1/§19.2). This endpoint therefore never touches credits.
    """

    outcome: Literal["kept", "rejected"]
    reason: RejectionReason | None = None
    note: str | None = Field(default=None, max_length=280)


class TryOnFeedbackResponse(BaseModel):
    result_id: str
    outcome: str
    recorded: bool
    learned: str | None = None
    profile: StyleMemoryProfile | None = None


class StyleMemorySignalOut(BaseModel):
    """One raw signal, for the "why does WTM think this?" view."""

    id: str
    signal_type: str
    value: str | None = None
    source: str
    created_at: datetime

File: app/lib/data/models/style_memory.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'style_memory.freezed.dart';
part 'style_memory.g.dart';

/// A structured reason a user gives for "Not me" (RETENTION spec §12.1).
///
/// Only three of these are TASTE. `identityIssue`, `garmentIssue` and
/// `bodyProportionIssue` are complaints about the render itself — the server
/// records them but deliberately does not learn a preference from them, so a
/// pipeline failure never becomes "you dislike navy".
enum RejectionReason {
  identityIssue('identity_issue'),
  garmentIssue('garment_issue'),
  notMyStyle('not_my_style'),
  bodyProportionIssue('body_proportion_issue'),
  colorIssue('color_issue'),
  occasionMismatch('occasion_mismatch'),
  other('other');

  const RejectionReason(this.wire);

  /// The value the API expects. Never send `name`.
  final String wire;
}

/// One thing WTM believes about the user, with the evidence behind it.
@freezed
abstract class PreferenceItem with _$PreferenceItem {
  const factory PreferenceItem({
    required String value,
    @Default(0) double weight,

    /// 0..1. Below [PreferenceItem.stateThreshold] the UI must phrase this as
    /// a hunch, never a fact (§12.3).
    @Default(0) double confidence,

    /// `stated` = the user told us. `inferred` = we noticed it.
    @Default('inferred') String source,
    @JsonKey(name: 'updated_at') String? updatedAt,
  }) = _PreferenceItem;

  const PreferenceItem._();

  factory PreferenceItem.fromJson(Map<String, dynamic> json) =>
      _$PreferenceItemFromJson(json);

  /// Mirrors the server's `_STATE_THRESHOLD`. Kept in sync deliberately: the
  /// server decides whether to WRITE a summary, the app decides whether to
  /// show a single preference as settled.
  static const stateThreshold = 0.35;

  /// True when the user said this outright, so it may be stated as fact.
  bool get isStated => source == 'stated';

  /// True when this is confident enough to show without hedging.
  bool get isConfident => isStated || confidence >= stateThreshold;
}

/// Everything WTM has learned about one user's taste.
@freezed
abstract class StyleMemoryProfile with _$StyleMemoryProfile {
  const factory StyleMemoryProfile({
    @Default(1) int version,
    @Default(0) double confidence,
    @JsonKey(name: 'signal_count') @Default(0) int signalCount,
    @JsonKey(name: 'personalization_enabled')
    @Default(true)
    bool personalizationEnabled,

    /// One hedged sentence, or null when we do not know enough to say anything.
    @JsonKey(name: 'preference_summary') String? preferenceSummary,
    @JsonKey(name: 'preferred_colors')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> preferredColors,
    @JsonKey(name: 'avoided_colors')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> avoidedColors,
    @JsonKey(name: 'preferred_silhouettes')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> preferredSilhouettes,
    @JsonKey(name: 'avoided_silhouettes')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> avoidedSilhouettes,
    @JsonKey(name: 'preferred_aesthetics')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> preferredAesthetics,
    @JsonKey(name: 'preferred_occasions')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> preferredOccasions,
    @JsonKey(name: 'preferred_moods')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> preferredMoods,
    @JsonKey(name: 'fit_visual_preferences')
    @Default(<PreferenceItem>[])
    List<PreferenceItem> fitVisualPreferences,
  }) = _StyleMemoryProfile;

  const StyleMemoryProfile._();

  factory StyleMemoryProfile.fromJson(Map<String, dynamic> json) =>
      _$StyleMemoryProfileFromJson(json);

  /// True when nothing has been learned yet — the honest empty state, and NOT
  /// an error.
  bool get isEmpty => signalCount == 0 && preferenceSummary == null;

  /// The facet keys shown on the "what WTM knows" screen, in reading order,
  /// paired with their entries.
  ///
  /// Deliberately returns the KEY, not a label: the key is what the correction
  /// endpoint expects, and the label is user-facing text that has to come from
  /// l10n (CLAUDE.md §4.3). A display string in a data model would be a string
  /// no translator ever sees.
  List<({String facet, List<PreferenceItem> items})> get facets => [
    (facet: 'preferred_colors', items: preferredColors),
    (facet: 'preferred_silhouettes', items: preferredSilhouettes),
    (facet: 'preferred_aesthetics', items: preferredAesthetics),
    (facet: 'preferred_occasions', items: preferredOccasions),
    (facet: 'preferred_moods', items: preferredMoods),
    (facet: 'avoided_colors', items: avoidedColors),
    (facet: 'avoided_silhouettes', items: avoidedSilhouettes),
  ];
}

/// The response to recording a signal — including the restrained "WTM
/// learned…" line, which is null whenever nothing actually changed.
@freezed
abstract class StyleMemorySignalResult with _$StyleMemorySignalResult {
  const factory StyleMemorySignalResult({
    @Default(false) bool recorded,
    String? learned,
    StyleMemoryProfile? profile,
  }) = _StyleMemorySignalResult;

  factory StyleMemorySignalResult.fromJson(Map<String, dynamic> json) =>
      _$StyleMemorySignalResultFromJson(json);
}

/// The result of submitting Keep it / Not me on a render.
@freezed
abstract class TryOnFeedbackResult with _$TryOnFeedbackResult {
  const factory TryOnFeedbackResult({
    @JsonKey(name: 'result_id') required String resultId,
    required String outcome,
    @Default(false) bool recorded,
    String? learned,
    StyleMemoryProfile? profile,
  }) = _TryOnFeedbackResult;

  factory TryOnFeedbackResult.fromJson(Map<String, dynamic> json) =>
      _$TryOnFeedbackResultFromJson(json);
}
